File: src/cliargsparser.cpp
#include "cliargsparser.h"

namespace Newsboat {

rust::Vec<cliargsparser::bridged::BytesVec> argv_to_rust_args(int argc, char* argv[])
{
	rust::Vec<cliargsparser::bridged::BytesVec> args;
	args.reserve(argc);
	for (int i = 0; i < argc; ++i) {
		cliargsparser::bridged::BytesVec arg;
		rust::Vec<uint8_t> data;
		for (auto j = 0; argv[i][j] != '\0'; ++j) {
			data.push_back(argv[i][j]);
		}
		arg.data = std::move(data);
		args.push_back(arg);
	}
	return args;
}

CliArgsParser::CliArgsParser(int argc, char* argv[])
	: rs_object(cliargsparser::bridged::create(argv_to_rust_args(argc, argv)))
{
}

bool CliArgsParser::do_import() const
{
	return Newsboat::cliargsparser::bridged::do_import(*rs_object);
}

bool CliArgsParser::do_export() const
{
	return Newsboat::cliargsparser::bridged::do_export(*rs_object);
}

bool CliArgsParser::export_as_opml2() const
{
	return Newsboat::cliargsparser::bridged::export_as_opml2(*rs_object);
}

bool CliArgsParser::do_vacuum() const
{
	return Newsboat::cliargsparser::bridged::do_vacuum(*rs_object);
}

bool CliArgsParser::do_cleanup() const
{
	return Newsboat::cliargsparser::bridged::do_cleanup(*rs_object);
}

std::string CliArgsParser::importfile() const
{
	return std::string(Newsboat::cliargsparser::bridged::importfile(*rs_object));
}

std::optional<std::string> CliArgsParser::readinfo_import_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::readinfo_import_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::readinfo_export_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::readinfo_export_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::string CliArgsParser::program_name() const
{
	return std::string(Newsboat::cliargsparser::bridged::program_name(*rs_object));
}

unsigned int CliArgsParser::show_version() const
{
	return Newsboat::cliargsparser::bridged::do_show_version(*rs_object);
}

bool CliArgsParser::silent() const
{
	return Newsboat::cliargsparser::bridged::silent(*rs_object);
}

bool CliArgsParser::using_nonstandard_configs() const
{
	return Newsboat::cliargsparser::bridged::using_nonstandard_configs(*rs_object);
}

std::optional<int> CliArgsParser::return_code() const
{
	rust::isize code = 0;
	if (Newsboat::cliargsparser::bridged::return_code(*rs_object, code)) {
		return static_cast<int>(code);
	}
	return std::nullopt;
}

std::string CliArgsParser::display_msg() const
{
	return std::string(Newsboat::cliargsparser::bridged::display_msg(*rs_object));
}

bool CliArgsParser::should_print_usage() const
{
	return Newsboat::cliargsparser::bridged::should_print_usage(*rs_object);
}

bool CliArgsParser::refresh_on_start() const
{
	return Newsboat::cliargsparser::bridged::refresh_on_start(*rs_object);
}

std::optional<std::string> CliArgsParser::url_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::url_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::lock_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::lock_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::cache_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::cache_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::config_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::config_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::queue_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::queue_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::search_history_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::search_history_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<std::string> CliArgsParser::cmdline_history_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::cmdline_history_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::vector<std::string> CliArgsParser::cmds_to_execute()
const
{
	const auto rs_cmds = Newsboat::cliargsparser::bridged::cmds_to_execute(
			*rs_object);

	std::vector<std::string> cmds;
	for (const auto& cmd : rs_cmds) {
		cmds.push_back(std::string(cmd));
	}
	return cmds;
}

std::optional<std::string> CliArgsParser::log_file() const
{
	rust::String path;
	if (Newsboat::cliargsparser::bridged::log_file(*rs_object, path)) {
		return std::string(path);
	}
	return std::nullopt;
}

std::optional<Level> CliArgsParser::log_level() const
{
	std::int8_t level;
	if (Newsboat::cliargsparser::bridged::log_level(*rs_object, level)) {
		return static_cast<Level>(level);
	}
	return std::nullopt;
}

const cliargsparser::bridged::CliArgsParser& CliArgsParser::get_rust_ref() const
{
	return *rs_object;
}

} // namespace Newsboat

File: rust/libnewsboat-ffi/src/cliargsparser.rs
use cxx::{type_id, ExternType};

use libNewsboat::cliargsparser;
use std::ffi::OsString;
use std::os::unix::ffi::OsStringExt;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
pub struct CliArgsParser(pub cliargsparser::CliArgsParser);

unsafe impl ExternType for CliArgsParser {
    type Id = type_id!("Newsboat::cliargsparser::bridged::CliArgsParser");
    type Kind = cxx::kind::Opaque;
}

#[cxx::bridge(namespace = "Newsboat::cliargsparser::bridged")]
mod bridged {
    /// Shared data structure between C++/Rust. We need to do this, because
    /// Vec<Vec<T>> is an unsupported CXX type.
    struct BytesVec {
        pub data: Vec<u8>,
    }

    extern "Rust" {
        type CliArgsParser;

        fn create(argv: Vec<BytesVec>) -> Box<CliArgsParser>;

        fn do_import(cliargsparser: &CliArgsParser) -> bool;
        fn do_export(cliargsparser: &CliArgsParser) -> bool;
        fn export_as_opml2(cliargsparser: &CliArgsParser) -> bool;
        fn do_vacuum(cliargsparser: &CliArgsParser) -> bool;
        fn do_cleanup(cliargsparser: &CliArgsParser) -> bool;
        fn do_show_version(cliargsparser: &CliArgsParser) -> u64;
        fn silent(cliargsparser: &CliArgsParser) -> bool;
        fn using_nonstandard_configs(cliargsparser: &CliArgsParser) -> bool;
        fn should_print_usage(cliargsparser: &CliArgsParser) -> bool;
        fn refresh_on_start(cliargsparser: &CliArgsParser) -> bool;

        fn importfile(cliargsparser: &CliArgsParser) -> String;
        fn program_name(cliargsparser: &CliArgsParser) -> String;
        fn display_msg(cliargsparser: &CliArgsParser) -> String;

        fn return_code(cliargsparser: &CliArgsParser, value: &mut isize) -> bool;

        fn readinfo_import_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn readinfo_export_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn url_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn lock_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn cache_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn config_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn queue_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn search_history_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn cmdline_history_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;
        fn log_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool;

        fn cmds_to_execute(cliargsparser: &CliArgsParser) -> Vec<String>;

        fn log_level(cliargsparser: &CliArgsParser, level: &mut i8) -> bool;
    }
}

fn create(argv: Vec<bridged::BytesVec>) -> Box<CliArgsParser> {
    let os_str_argv: Vec<OsString> = argv
        .into_iter()
        .map(|arg_bytes| OsString::from_vec(arg_bytes.data))
        .collect();
    Box::new(CliArgsParser(cliargsparser::CliArgsParser::new(
        os_str_argv,
    )))
}

fn do_import(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.importfile.is_some()
}

fn do_export(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.do_export
}

fn export_as_opml2(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.export_as_opml2
}

fn do_vacuum(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.do_vacuum
}

fn do_cleanup(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.do_cleanup
}

fn do_show_version(cliargsparser: &CliArgsParser) -> u64 {
    cliargsparser.0.show_version as u64
}

fn silent(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.silent
}

fn using_nonstandard_configs(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.using_nonstandard_configs()
}

fn should_print_usage(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.should_print_usage
}

fn refresh_on_start(cliargsparser: &CliArgsParser) -> bool {
    cliargsparser.0.refresh_on_start
}

fn importfile(cliargsparser: &CliArgsParser) -> String {
    match &cliargsparser.0.importfile {
        Some(path) => path.to_string_lossy().to_string(),
        None => String::new(),
    }
}

fn program_name(cliargsparser: &CliArgsParser) -> String {
    cliargsparser.0.program_name.to_string()
}

fn display_msg(cliargsparser: &CliArgsParser) -> String {
    cliargsparser.0.display_msg.to_string()
}

fn return_code(cliargsparser: &CliArgsParser, value: &mut isize) -> bool {
    match cliargsparser.0.return_code {
        Some(code) => {
            *value = code as isize;
            true
        }
        None => false,
    }
}

fn readinfo_import_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.readinfo_import_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn readinfo_export_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.readinfo_export_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn url_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.url_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn lock_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.lock_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn cache_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.cache_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn config_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.config_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn queue_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.queue_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn search_history_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.search_history_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn cmdline_history_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.cmdline_history_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn log_file(cliargsparser: &CliArgsParser, path: &mut String) -> bool {
    match &cliargsparser.0.log_file {
        Some(p) => {
            *path = p.to_string_lossy().to_string();
            true
        }
        None => false,
    }
}

fn cmds_to_execute(cliargsparser: &CliArgsParser) -> Vec<String> {
    cliargsparser.0.cmds_to_execute.to_owned()
}

fn log_level(cliargsparser: &CliArgsParser, level: &mut i8) -> bool {
    match cliargsparser.0.log_level {
        Some(l) => {
            *level = l as i8;
            true
        }
        None => false,
    }
}
